release-20.1: kv: unset CanForwardReadTimestamp flag on batches that spans ranges

nvb opened the pull request:
Backport 2/4 commits from #50633.

/cc @cockroachdb/release

---

Fixes #50202.

In #50202, we saw that we would accidentally allow batches to be split on range boundaries and continue to carry the `CanForwardReadTimestamp` flag. This can lead to serializability violations under very specific conditions:
1. the first operation that the transaction performs is a batch of locking read requests (due to implicit or explicit SFU)
2. this batch of locking reads spans multiple ranges
3. this batch of locking reads is issued in parallel by the DistSender
4. this locking read hits contention and is bumped on at least one of the ranges due to a WriteTooOld error
5. an unreplicated lock from one of the non-refreshed sub-batches is lost during a lease transfer.

It turns out that the `kv/contention` roachtest meets these requirements perfectly when implicit SFU support is added to UPSERT statements: #50180. It creates a tremendous amount of contention and issues a batch of locking ScanRequests during a LookupJoin as its first operation. This materializes as ConditionFailedErrors (which should be impossible) in the CPuts that the UPSERT issues to maintain the table's secondary index.

This PR fixes this bug by ensuring that if a batch is going to be split across ranges and any of its requests would need to refresh on read timestamp bumps, it does not have its CanForwardReadTimestamp flag set. It would be incorrect to allow part of a batch to perform a server-side refresh if another part of the batch might have returned a different result at the higher timestamp, which is a fancy way of saying that it needs to refresh because it is using optimistic locking. Such behavior could cause a transaction to observe an inconsistent snapshot and violate serializability.

It then adds support for locking scans to kvnemesis, which would have caught to bug fairly easily.

Finally, it fixes a KV API UX issue around locking scans and retry errors. Before this change, it was possible for a non-transactional locking scan (which itself doesn't make much sense) to hit a WriteTooOld  retry error. This was caused by eager propagation of WriteTooOld errors from MVCC when FailOnMoreRecent was enabled for an MVCCScan. I'd appreciate if @itsbilal could give that last commit a review.

Release note (bug fix): fix a rare bug where a multi-Range SELECT FOR UPDATE statement containing an IN clause could fail to observe a consistent snapshot and violate serializability.

nvb committed:
kv: add log event for unreplicated lock acquisition

nvb committed:
kv/kvclient: unset CanForwardReadTimestamp flag on batch that spans ranges

Fixes #50202.

In #50202, we saw that we would accidentally allow batches to be split
on range boundaries and continue to carry the `CanForwardReadTimestamp`
flag. This can lead to serializability violations under very specific
conditions:
1. the first operation that the transaction performs is a batch of
   locking read requests (due to implicit or explicit SFU)
2. this batch of locking reads spans multiple ranges
3. this batch of locking reads is issued in parallel by the DistSender
4. this locking read hits contention and is bumped on at least one of
   the ranges due to a WriteTooOld error
5. an unreplicated lock from one of the non-refreshed sub-batches is
   lost during a lease transfer.

It turns out that the `kv/contention` roachtest meets these requirements
perfectly when implicit SFU support is added to UPSERT statements: #50180.
It creates a tremendous amount of contention and issues a batch
of locking ScanRequests during a LookupJoin as its first operation. This
materializes as ConditionFailedErrors (which should be impossible) in
the CPuts that the UPSERT issues to maintain the table's secondary
index.

This commit fixes this bug by ensuring that if a batch is going to be
split across ranges and any of its requests would need to refresh on
read timestamp bumps, it does not have its CanForwardReadTimestamp flag
set. It would be incorrect to allow part of a batch to perform a
server-side refresh if another part of the batch might have returned a
different result at the higher timestamp, which is a fancy way of saying
that it needs to refresh because it is using optimistic locking. Such
behavior could cause a transaction to observe an inconsistent snapshot
and violate serializability.

Release note (bug fix): fix a rare bug where a multi-Range SELECT FOR
UPDATE statement containing an IN clause could fail to observe a
consistent snapshot and violate serializability.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50816)
<!-- Reviewable:end -->

nvb committed:
kv/kvclient: avoid letting BatchRequests escape to heap

The first one in `divideAndSendBatchToRanges` was caused by #50633.

While here, avoid the allocation in `divideAndSendParallelCommit`.

nvanbenschoten commented:
I'm going to add https://github.com/cockroachdb/cockroach/pull/50819 to this once it's merged.

andreimatei commented:
LGTM

On Tue, Jun 30, 2020 at 9:34 AM Nathan VanBenschoten <
notifications@github.com> wrote:

> I'm going to add #50819
> <https://github.com/cockroachdb/cockroach/pull/50819> to this once it's
> merged.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/50816#issuecomment-651794110>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC4C4OHMV3UFOUUVHUXBCTRZHSVDANCNFSM4OMHKX5Q>
> .
>

nvanbenschoten commented:
Added https://github.com/cockroachdb/cockroach/pull/50819 to this.

andreimatei commented:
LGTM

On Wed, Jul 1, 2020 at 12:18 PM Nathan VanBenschoten <
notifications@github.com> wrote:

> Added #50819 <https://github.com/cockroachdb/cockroach/pull/50819> to
> this.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/50816#issuecomment-652515842>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC4C4NOWVXVWARGCRB5E4TRZNOWNANCNFSM4OMHKX5Q>
> .
>
